src/lib/int.ts

```diff
@@ -0,0 +1,392 @@
+  }
+
+  divMod(y: UInt32 | number): [UInt32, UInt32] {
+    let x = this.value;
+    let y_ = argToField('UInt32.div', y);
+
+    if (this.value.isConstant() && y_.isConstant()) {
+      let xn = BigInt(x.toString());
+      let yn = BigInt(y_.toString());
+      let q = xn / yn;
+      let r = xn - q * yn;
+      return [
+        new UInt32(new Field(q.toString())),
+        new UInt32(new Field(r.toString())),
+      ];
+    }
+
+    y_ = y_.seal();
+
+    let q = Circuit.witness(
+      Field,
+      () => new Field((BigInt(x.toString()) / BigInt(y_.toString())).toString())
+    );
+
+    q.rangeCheckHelper(UInt32.NUM_BITS).assertEquals(q);
+
+    // TODO: Could be a bit more efficient
+    let r = x.sub(q.mul(y_)).seal();
+    r.rangeCheckHelper(UInt32.NUM_BITS).assertEquals(r);
+
+    let r_ = new UInt32(r);
+    let q_ = new UInt32(q);
+
+    r_.assertLt(new UInt32(y_));
+
+    return [q_, r_];
+  }
+
+  div(y: UInt32 | number): UInt32 {
+    const dm = this.divMod(y);
+    return dm[0];
+  }
+
+  mod(y: UInt32 | number): UInt32 {
+    return this.divMod(y)[1];
+  }
+
+  mul(y: UInt32 | number): UInt32 {
+    let z = this.value.mul(argToField('UInt32.mul', y));
+    z.rangeCheckHelper(UInt32.NUM_BITS).assertEquals(z);
+    return new UInt32(z);
+  }
+
+  add(y: UInt32 | number): UInt32 {
+    let z = this.value.add(argToField('UInt32.add', y));
+    z.rangeCheckHelper(UInt32.NUM_BITS).assertEquals(z);
+    return new UInt32(z);
+  }
+
+  sub(y: UInt32 | number): UInt32 {
+    let z = this.value.sub(argToField('UInt32.sub', y));
+    z.rangeCheckHelper(UInt32.NUM_BITS).assertEquals(z);
+    return new UInt32(z);
+  }
+
+  lte(y: UInt32): Bool {
+    let xMinusY = this.value.sub(argToField('UInt32.lte', y)).seal();
+    let xMinusYFits = xMinusY.rangeCheckHelper(UInt32.NUM_BITS).equals(xMinusY);
+    let yMinusXFits = xMinusY
+      .rangeCheckHelper(UInt32.NUM_BITS)
+      .equals(xMinusY.neg());
+    xMinusYFits.or(yMinusXFits).assertEquals(true);
+    // x <= y if y - x fits in 32 bits
+    return yMinusXFits;
+  }
+
+  assertLte(y: UInt32) {
+    let yMinusX = argToField('UInt32.lt', y).sub(this.value).seal();
+    yMinusX.rangeCheckHelper(UInt32.NUM_BITS).assertEquals(yMinusX);
+  }
+
+  lt(y: UInt32): Bool {
+    return this.lte(y).and(this.value.equals(y.value).not());
+  }
+
+  assertLt(y: UInt32) {
+    this.lt(y).assertEquals(true);
+  }
+
+  gt(y: UInt32): Bool {
+    return y.lt(this);
+  }
+
+  assertGt(y: UInt32) {
+    y.assertLt(this);
+  }
+}
+
+class Sgn extends CircuitValue {
+  // +/- 1
+  @prop value: Field;
+
+  static check(x: Sgn) {
+    let x_ = x.value.seal();
+    x_.mul(x_).assertEquals(Field.one);
+  }
+
+  constructor(value: Field) {
+    super();
+    this.value = value;
+  }
+
+  static get Pos() {
+    return new Sgn(Field.one);
+  }
+  static get Neg() {
+    return new Sgn(Field.one.neg());
+  }
+}
+
+export class Int64 {
+  // In the range [-2^63, 2^63 - 1]
+  @prop value: Field;
```

Comment:
`Int64` doesn't really seem implemented yet - it looks like it's just a wrapper around a Field, without anything constraining it to 64 bits.
This makes me a bit uneasy because we use `Int64` as the type for the delta in transactions

Reply:
Made an issue